File: GENERATOR_TESTE_PY/test_generator.py
from random import randint
from math import ceil
from random import sample
import time

def random_generator(run):
    out = open("data.in", "a")

    fuel = 700  # maximul de km pe care se poate deplasa masina fara sa opreasca
    max_range = 1000000  # distanta maxima la care se poate deplase masina

    for case_number in range(run):
        # masina se poate deplasa pe o distanta de cel putin 400 km si cel mult FUEL km
        m = randint(400, fuel)
        # destinatia se alfa la cel putin 1000 km departare si la maxim max_range km departare
        destination = randint(600 + m, max_range)
        # sunt cel putin n_must_fill benzinarii la care trebuie sa opreasca
        # si cel mult destination/10 + n_must_fill benzinarii optionale
        n_must_fill = ceil(destination / m)
        n = n_must_fill + randint(0, destination // 10)

        arr1 = [0]
        r_value = 0
        while r_value < destination:
            r_value = r_value + m - randint(1, 100)
            arr1.append(r_value)

        arr2 = list(sample(range(1, destination - 1), n - 1))
        arr2.append(destination)

        arr = list(set(arr1) | set(arr2))

        arr.sort()

        n = len(arr) - 2
        # START TESTARE

        for iterator in range(1, n + 2):
            assert (arr[iterator] - arr[iterator - 1] <= m)
            assert arr[iterator] != arr[iterator - 1]
            assert arr[iterator] > arr[iterator - 1]

        # STOP TESTARE

        out.write("%d\n" % n)
        out.write("%d\n" % m)

        for iterator in arr:
            out.write("%s\n" % iterator)

    out.close()

File: GENERATOR_TESTE/test_generator.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <math.h>
#include "utility.h"
#include "sorting_algorithms.h"

#define FUEL 700  // maximul de km pe care se poate deplasa masina fara sa opreasca

#define MAX 1000000   //distanta maxima la care se poate deplase masina


void random_generator(int no_test)
{
    FILE* out = fopen("data.in", "a");

    srand(time(NULL));
    int n; //numarul de benzinarii
    int m;  //numarul de km pe care se poate deplasa masina fara oprire
    int n_must_fill;
    int iterator;
    int destination;
    int case_numeber;

    for(case_numeber=0; case_numeber<no_test; case_numeber++)
    {
        /* masina se poate deplasa pe o distanta de cel putin 400 km
        si cel mult FUEL km*/
        m = 400 + rand_range(FUEL- 400 + 1);

        /* destinatia se alfa la cel putin 1000 km departare
            si la maxim MAX km departare*/
        destination  = 600 + m + rand_range(MAX - (600 + m) + 1);

        /* sunt cel putin n_must_fill benzinarii la care trebuie sa opreasca
           si cel mult destination/10 + n_must_fill benzinarii optionale*/
        n_must_fill = ceil((double)destination/m);

        n = n_must_fill + rand_range(destination/10);

        int* arr = (int*)calloc(n+2,sizeof(int));

        arr[0] = 0;
        arr[n+1] = destination;

        int* is_used= (int*)calloc(destination, sizeof(int)); //memoram variabilele ce au fost deja folosite
        is_used[arr[0]]=1;

        int r_value;

        for( iterator =1 ;   iterator <= n; iterator++)
        {

            r_value = arr[iterator-1] + m - rand()%100; // cream benzinariile sigure la care trebuie sa opreasca soferul

            if(r_value < destination && is_used[r_value] == 0)
            {
                arr[iterator] = r_value;
                is_used[r_value] = 1;

            }
            else
            {
                while(1)
                {
                    r_value = 1 +  rand_range(destination-1);   // daca am generat toate benzinariile sigure atunci generam benzinarii optionale
                    if(!is_used[r_value])      // daca valoarea nu a mai fost memorata atunci o folosim
                    {
                        break;
                    }
                }

                arr[iterator] = r_value;
                is_used[r_value] = 1;
            }
        }

        mergeSort(arr, 0, n+1);

        n = n - 1;

        test(n, m, arr); // Testam daca datele sunt relevante

        fprintf(out,"%d\n", n);

        fprintf(out,"%d\n", m);

        for(iterator=0; iterator<n+2; iterator++)
        {
            fprintf(out,"%d\n", arr[iterator]);
        }

        free(is_used);
        free(arr);

    }

    fclose(out);

}
